internal/info/info.go

```diff
@@ -54,43 +69,89 @@ type StackTrace struct {
 }
 
 var (
-	Version      = "v0.0.1"
-	GitCommit    = "master"
+	// injected from build script
+	Version           = "v0.0.1"
+	GitCommit         = "master"
+	BuildTime         = ""
+	GoVersion         string
+	GoOS              string
+	GoArch            string
+	CGOEnabled        string
+	NGTVersion        string
+	BuildCPUInfoFlags string
+
 	Organization = "vdaas"
 	Repository   = "vald"
-	BuildTime    = ""
 
-	GoVersion  string
-	GoOS       string
-	GoArch     string
-	CGOEnabled string
+	reps = strings.NewReplacer("_", " ", ",omitempty", "")
 
-	NGTVersion string
+	once         sync.Once
+	infoProvider Info
+)
 
-	BuildCPUInfoFlags string
+// Init initializes Detail object only once.
+func Init(name string) error {
+	i, err := New(WithServerName(name))
+	if err != nil {
+		return err
+	}
+	once.Do(func() {
+		infoProvider = i
+	})
+	return nil
+}
 
-	reps = strings.NewReplacer("_", " ", ",omitempty", "")
+// New initializes and returns the info object or any error occurred.
+func New(opts ...Option) (Info, error) {
+	i := &info{
+		detail: Detail{
+			Version:           Version,
+			GitCommit:         GitCommit,
+			BuildTime:         BuildTime,
+			GoVersion:         GoVersion,
+			GoOS:              GoOS,
+			GoArch:            GoArch,
+			CGOEnabled:        CGOEnabled,
+			NGTVersion:        NGTVersion,
+			BuildCPUInfoFlags: strings.Split(strings.TrimSpace(BuildCPUInfoFlags), " "),
+		},
+	}
 
-	once sync.Once
+	for _, opt := range append(defaultOpts, opts...) {
+		if err := opt(i); err != nil {
+			werr := errors.ErrOptionFailed(err, reflect.ValueOf(opt))
 
-	detail Detail
-)
+			e := new(errors.ErrCriticalOption)
+			if errors.As(err, &e) {
+				log.Error(werr)
+				return nil, werr
+			}
+			log.Warn(werr)
+		}
+	}
+
+	i.prepare()
+
+	return i, nil
+}
 
 // String calls String method of global detail object.
 func String() string {
-	return detail.String()
+	return infoProvider.String()
 }
 
 // Get calls Get method of global detail object.
 func Get() Detail {
-	return detail.Get()
+	return infoProvider.Get()
 }
 
 // String returns summary of Detail object.
-func (d Detail) String() string {
-	if len(d.StackTrace) == 0 {
-		d = d.Get()
+func (i info) String() string {
```

Comment:
same
https://github.com/vdaas/vald/pull/862/files#r549204368

Reply:
replied on that thread.